File: data/redis/code/chapter9/python/message_queue.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Redis消息队列实现
支持List队列、Pub/Sub、Stream等模式
"""

import redis
import json
import time
import threading
import random
from typing import Any, Callable, List, Optional
from concurrent.futures import ThreadPoolExecutor
from threading import Event


class ListMessageQueue:
    """List队列实现（简单的FIFO队列）"""
    
    def __init__(self, redis_client: redis.Redis, queue_name: str):
        self.redis = redis_client
        self.queue_name = queue_name
    
    def send_message(self, message: Any) -> bool:
        """发送消息"""
        try:
            serialized_message = json.dumps(message)
            self.redis.rpush(self.queue_name, serialized_message)
            return True
        except Exception as e:
            print(f"消息发送失败: {e}")
            return False
    
    def receive_message(self, timeout: int = 30) -> Optional[Any]:
        """接收消息（阻塞）"""
        try:
            # 阻塞弹出消息
            result = self.redis.blpop(self.queue_name, timeout=timeout)
            
            if result is None:
                return None
            
            _, message = result
            return json.loads(message)
        except Exception as e:
            print(f"消息接收失败: {e}")
            return None
    
    def batch_send_messages(self, messages: List[Any]) -> bool:
        """批量发送消息"""
        try:
            serialized_messages = [json.dumps(msg) for msg in messages]
            self.redis.rpush(self.queue_name, *serialized_messages)
            return True
        except Exception as e:
            print(f"批量消息发送失败: {e}")
            return False


class PubSubMessageQueue:
    """Pub/Sub消息队列实现"""
    
    def __init__(self, redis_client: redis.Redis, channel_name: str):
        self.redis = redis_client
        self.channel_name = channel_name
        self.pubsub = self.redis.pubsub()
        self.running = False
        self.thread = None
    
    def publish_message(self, message: Any) -> bool:
        """发布消息"""
        try:
            serialized_message = json.dumps(message)
            self.redis.publish(self.channel_name, serialized_message)
            return True
        except Exception as e:
            print(f"消息发布失败: {e}")
            return False
    
    def subscribe(self, message_handler: Callable[[Any], None]):
        """订阅消息"""
        def message_listener():
            self.pubsub.subscribe(self.channel_name)
            
            for message in self.pubsub.listen():
                if not self.running:
                    break
                
                if message['type'] == 'message':
                    try:
                        data = json.loads(message['data'])
                        message_handler(data)
                    except Exception as e:
                        print(f"消息处理失败: {e}")
        
        self.running = True
        self.thread = threading.Thread(target=message_listener)
        self.thread.daemon = True
        self.thread.start()
    
    def unsubscribe(self):
        """取消订阅"""
        self.running = False
        if self.thread:
            self.pubsub.unsubscribe(self.channel_name)
            self.thread.join(timeout=5)


class StreamMessageQueue:
    """Stream消息队列实现（Redis 5.0+）"""
    
    def __init__(self, redis_client: redis.Redis, stream_name: str):
        self.redis = redis_client
        self.stream_name = stream_name
    
    def send_message(self, message: Any) -> Optional[str]:
        """发送消息"""
        try:
            serialized_message = json.dumps(message)
            
            message_data = {
                'data': serialized_message,
                'timestamp': str(int(time.time()))
            }
            
            # 使用自动生成的ID
            message_id = self.redis.xadd(self.stream_name, message_data)
            return message_id
        except Exception as e:
            print(f"Stream消息发送失败: {e}")
            return None
    
    def read_messages(self, count: int = 10, start_id: str = '0') -> List[Any]:
        """读取消息"""
        try:
            # 从指定位置开始读取
            results = self.redis.xread(
                {self.stream_name: start_id},
                count=count,
                block=0
            )
            
            messages = []
            
            if results:
                for stream_name, entries in results:
                    for entry_id, entry_data in entries:
                        try:
                            message_data = json.loads(entry_data['data'])
                            messages.append({
                                'id': entry_id,
                                'data': message_data,
                                'timestamp': entry_data.get('timestamp', '')
                            })
                        except Exception as e:
                            print(f"消息解析失败: {e}")
            
            return messages
        except Exception as e:
            print(f"Stream消息读取失败: {e}")
            return []
    
    def create_consumer_group(self, group_name: str, start_id: str = '0') -> bool:
        """创建消费者组"""
        try:
            self.redis.xgroup_create(self.stream_name, group_name, start_id, mkstream=True)
            return True
        except Exception as e:
            # 如果组已存在，忽略错误
            if "BUSYGROUP" in str(e):
                return True
            print(f"创建消费者组失败: {e}")
            return False
    
    def read_messages_by_group(self, group_name: str, consumer_name: str, 
                              count: int = 10) -> List[Any]:
        """消费者组读取消息"""
        try:
            # 消费者组读取
            results = self.redis.xreadgroup(
                group_name, 
                consumer_name,
                {self.stream_name: '>'},
                count=count,
                block=5000
            )
            
            messages = []
            
            if results:
                for stream_name, entries in results:
                    for entry_id, entry_data in entries:
                        try:
                            message_data = json.loads(entry_data['data'])
                            messages.append({
                                'id': entry_id,
                                'data': message_data,
                                'timestamp': entry_data.get('timestamp', '')
                            })
                            
                            # 确认消息处理
                            self.redis.xack(self.stream_name, group_name, entry_id)
                        except Exception as e:
                            print(f"消息解析失败: {e}")
            
            return messages
        except Exception as e:
            print(f"消费者组消息读取失败: {e}")
            return []


class DelayedMessageQueue:
    """延迟消息队列实现"""
    
    def __init__(self, redis_client: redis.Redis, queue_name: str):
        self.redis = redis_client
        self.queue_name = queue_name
        self.delayed_set = f"{queue_name}:delayed"
        self.running = False
        self.thread = None
    
    def send_delayed_message(self, message: Any, delay_seconds: int) -> bool:
        """发送延迟消息"""
        try:
            serialized_message = json.dumps(message)
            deliver_time = int(time.time()) + delay_seconds
            
            # 使用有序集合存储延迟消息
            self.redis.zadd(self.delayed_set, {serialized_message: deliver_time})
            
            # 启动检查线程（如果未运行）
            if not self.running:
                self.start_delivery_check()
            
            return True
        except Exception as e:
            print(f"延迟消息发送失败: {e}")
            return False
    
    def start_delivery_check(self):
        """启动投递检查线程"""
        def delivery_checker():
            while self.running:
                try:
                    current_time = int(time.time())
                    
                    # 获取所有到期的消息
                    expired_messages = self.redis.zrangebyscore(
                        self.delayed_set, 0, current_time
                    )
                    
                    if expired_messages:
                        # 投递到实际队列
                        self.redis.rpush(self.queue_name, *expired_messages)
                        
                        # 从延迟队列中移除
                        self.redis.zremrangebyscore(self.delayed_set, 0, current_time)
                    
                    # 每秒检查一次
                    time.sleep(1)
                except Exception as e:
                    print(f"投递检查失败: {e}")
                    time.sleep(5)  # 错误时等待更长时间
        
        self.running = True
        self.thread = threading.Thread(target=delivery_checker)
        self.thread.daemon = True
        self.thread.start()
    
    def receive_message(self, timeout: int = 30) -> Optional[Any]:
        """接收消息"""
        try:
            result = self.redis.blpop(self.queue_name, timeout=timeout)
            
            if result is None:
                return None
            
            _, message = result
            return json.loads(message)
        except Exception as e:
            print(f"消息接收失败: {e}")
            return None
    
    def stop(self):
        """停止队列"""
        self.running = False
        if self.thread:
            self.thread.join(timeout=5)


def test_message_queues():
    """测试消息队列功能"""
    # 创建Redis连接
    r = redis.Redis(host='localhost', port=6379, db=0, decode_responses=True)
    
    print("=== 测试List消息队列 ===")
    list_queue = ListMessageQueue(r, "test:list:queue")
    
    # 发送测试消息
    test_message = {"type": "test", "content": "Hello List Queue"}
    list_queue.send_message(test_message)
    
    # 接收消息
    received_message = list_queue.receive_message(timeout=5)
    print(f"List队列接收消息: {received_message}")
    
    print("\n=== 测试Pub/Sub消息队列 ===")
    pubsub_queue = PubSubMessageQueue(r, "test:pubsub:channel")
    
    # 使用事件等待消息接收
    message_received = Event()
    received_messages = []
    
    def message_handler(message):
        received_messages.append(message)
        print(f"Pub/Sub队列接收消息: {message}")
        message_received.set()
    
    # 订阅消息
    pubsub_queue.subscribe(message_handler)
    
    # 等待订阅建立
    time.sleep(0.1)
    
    # 发布消息
    pubsub_message = {"type": "test", "content": "Hello Pub/Sub"}
    pubsub_queue.publish_message(pubsub_message)
    
    # 等待消息接收
    message_received.wait(timeout=5)
    pubsub_queue.unsubscribe()
    
    print("\n=== 测试Stream消息队列 ===")
    stream_queue = StreamMessageQueue(r, "test:stream:queue")
    
    # 发送消息
    stream_message = {"type": "test", "content": "Hello Stream"}
    message_id = stream_queue.send_message(stream_message)
    print(f"Stream队列发送消息，ID: {message_id}")
    
    # 读取消息
    stream_messages = stream_queue.read_messages(count=10)
    print(f"Stream队列读取消息: {stream_messages}")
    
    # 测试消费者组
    print("\n=== 测试Stream消费者组 ===")
    stream_queue.create_consumer_group("test_group", "0")
    
    # 消费者组读取
    group_messages = stream_queue.read_messages_by_group(
        "test_group", "consumer1", count=10
    )
    print(f"消费者组读取消息: {group_messages}")
    
    print("\n=== 测试延迟消息队列 ===")
    delayed_queue = DelayedMessageQueue(r, "test:delayed:queue")
    
    # 发送延迟消息（2秒后投递）
    delayed_message = {"type": "test", "content": "Hello Delayed Queue"}
    delayed_queue.send_delayed_message(delayed_message, 2)
    
    # 立即尝试接收（应该收不到）
    immediate_message = delayed_queue.receive_message(timeout=1)
    print(f"立即接收延迟消息: {immediate_message}")
    
    # 等待2秒后接收
    time.sleep(2.5)
    delayed_received_message = delayed_queue.receive_message(timeout=1)
    print(f"延迟后接收消息: {delayed_received_message}")
    
    delayed_queue.stop()
    
    print("\n消息队列测试完成!")


if __name__ == "__main__":
    test_message_queues()

File: data-analystics/redis/code/chapter9/java/src/main/java/com/redis/tutorial/MessageQueue.java
package com.redis.tutorial;

import redis.clients.jedis.Jedis;
import redis.clients.jedis.StreamEntry;
import redis.clients.jedis.StreamEntryID;
import redis.clients.jedis.params.XAddParams;
import redis.clients.jedis.params.XReadParams;
import redis.clients.jedis.params.XReadGroupParams;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.*;
import java.util.function.Consumer;

/**
 * Redis消息队列实现
 * 支持List队列、Pub/Sub、Stream等模式
 */
public class MessageQueue {
    
    private static final Logger logger = LoggerFactory.getLogger(MessageQueue.class);
    private static final ObjectMapper objectMapper = new ObjectMapper();
    
    private final Jedis jedis;
    
    public MessageQueue(Jedis jedis) {
        this.jedis = jedis;
    }
}

/**
 * List队列实现（简单的FIFO队列）
 */
class ListMessageQueue {
    
    private final Jedis jedis;
    private final String queueName;
    
    public ListMessageQueue(Jedis jedis, String queueName) {
        this.jedis = jedis;
        this.queueName = queueName;
    }
    
    /**
     * 发送消息
     */
    public <T> boolean sendMessage(T message) {
        try {
            String serializedMessage = objectMapper.writeValueAsString(message);
            jedis.rpush(queueName, serializedMessage);
            return true;
        } catch (JsonProcessingException e) {
            logger.error("消息序列化失败: {}", e.getMessage());
            return false;
        }
    }
    
    /**
     * 接收消息（阻塞）
     */
    public <T> T receiveMessage(Class<T> clazz, int timeoutSeconds) {
        List<String> messages = jedis.blpop(timeoutSeconds, queueName);
        
        if (messages != null && messages.size() >= 2) {
            String message = messages.get(1);
            try {
                return objectMapper.readValue(message, clazz);
            } catch (JsonProcessingException e) {
                logger.error("消息反序列化失败: {}", e.getMessage());
                return null;
            }
        }
        
        return null;
    }
    
    /**
     * 批量发送消息
     */
    public <T> boolean batchSendMessages(List<T> messages) {
        try {
            String[] serializedMessages = messages.stream()
                    .map(message -> {
                        try {
                            return objectMapper.writeValueAsString(message);
                        } catch (JsonProcessingException e) {
                            logger.error("消息序列化失败: {}", e.getMessage());
                            return null;
                        }
                    })
                    .filter(Objects::nonNull)
                    .toArray(String[]::new);
            
            if (serializedMessages.length > 0) {
                jedis.rpush(queueName, serializedMessages);
                return true;
            }
            return false;
        } catch (Exception e) {
            logger.error("批量发送消息失败: {}", e.getMessage());
            return false;
        }
    }
}

/**
 * Pub/Sub消息队列实现
 */
class PubSubMessageQueue {
    
    private final Jedis jedis;
    private final String channelName;
    private final ExecutorService executorService;
    
    public PubSubMessageQueue(Jedis jedis, String channelName) {
        this.jedis = jedis;
        this.channelName = channelName;
        this.executorService = Executors.newCachedThreadPool();
    }
    
    /**
     * 发布消息
     */
    public <T> boolean publishMessage(T message) {
        try {
            String serializedMessage = objectMapper.writeValueAsString(message);
            jedis.publish(channelName, serializedMessage);
            return true;
        } catch (JsonProcessingException e) {
            logger.error("消息序列化失败: {}", e.getMessage());
            return false;
        }
    }
    
    /**
     * 订阅消息
     */
    public <T> void subscribeMessage(Class<T> clazz, Consumer<T> messageHandler) {
        executorService.submit(() -> {
            jedis.subscribe(new redis.clients.jedis.JedisPubSub() {
                @Override
                public void onMessage(String channel, String message) {
                    if (channelName.equals(channel)) {
                        try {
                            T deserializedMessage = objectMapper.readValue(message, clazz);
                            messageHandler.accept(deserializedMessage);
                        } catch (JsonProcessingException e) {
                            logger.error("消息反序列化失败: {}", e.getMessage());
                        }
                    }
                }
            }, channelName);
        });
    }
    
    /**
     * 关闭订阅
     */
    public void shutdown() {
        executorService.shutdown();
    }
}

/**
 * Stream消息队列实现（Redis 5.0+）
 */
class StreamMessageQueue {
    
    private final Jedis jedis;
    private final String streamName;
    
    public StreamMessageQueue(Jedis jedis, String streamName) {
        this.jedis = jedis;
        this.streamName = streamName;
    }
    
    /**
     * 发送消息
     */
    public <T> StreamEntryID sendMessage(T message) {
        try {
            String serializedMessage = objectMapper.writeValueAsString(message);
            
            Map<String, String> messageMap = new HashMap<>();
            messageMap.put("data", serializedMessage);
            messageMap.put("timestamp", String.valueOf(System.currentTimeMillis()));
            
            XAddParams params = XAddParams.xAddParams();
            return jedis.xadd(streamName, params, messageMap);
        } catch (JsonProcessingException e) {
            logger.error("消息序列化失败: {}", e.getMessage());
            return null;
        }
    }
    
    /**
     * 读取消息
     */
    public <T> List<T> readMessages(Class<T> clazz, int count) {
        return readMessages(clazz, count, "0");
    }
    
    /**
     * 从指定位置读取消息
     */
    public <T> List<T> readMessages(Class<T> clazz, int count, String startId) {
        Map<String, StreamEntryID> streams = new HashMap<>();
        streams.put(streamName, new StreamEntryID(startId));
        
        XReadParams params = XReadParams.xReadParams().count(count);
        
        List<Map.Entry<String, List<StreamEntry>>> results = 
                jedis.xread(params, streams);
        
        List<T> messages = new ArrayList<>();
        
        for (Map.Entry<String, List<StreamEntry>> result : results) {
            for (StreamEntry entry : result.getValue()) {
                try {
                    String messageData = entry.getFields().get("data");
                    T message = objectMapper.readValue(messageData, clazz);
                    messages.add(message);
                } catch (JsonProcessingException e) {
                    logger.error("消息反序列化失败: {}", e.getMessage());
                }
            }
        }
        
        return messages;
    }
    
    /**
     * 创建消费者组
     */
    public boolean createConsumerGroup(String groupName, String startId) {
        try {
            jedis.xgroupCreate(streamName, groupName, new StreamEntryID(startId), true);
            return true;
        } catch (Exception e) {
            logger.error("创建消费者组失败: {}", e.getMessage());
            return false;
        }
    }
    
    /**
     * 消费者组读取消息
     */
    public <T> List<T> readMessagesByGroup(Class<T> clazz, String groupName, 
                                          String consumerName, int count) {
        Map<String, StreamEntryID> streams = new HashMap<>();
        streams.put(streamName, StreamEntryID.UNRECEIVED_ENTRY);
        
        XReadGroupParams params = XReadGroupParams.xReadGroupParams()
                .count(count)
                .block(5000); // 5秒阻塞
        
        List<Map.Entry<String, List<StreamEntry>>> results = 
                jedis.xreadGroup(groupName, consumerName, params, streams);
        
        List<T> messages = new ArrayList<>();
        
        for (Map.Entry<String, List<StreamEntry>> result : results) {
            for (StreamEntry entry : result.getValue()) {
                try {
                    String messageData = entry.getFields().get("data");
                    T message = objectMapper.readValue(messageData, clazz);
                    messages.add(message);
                    
                    // 确认消息处理
                    jedis.xack(streamName, groupName, entry.getID());
                } catch (JsonProcessingException e) {
                    logger.error("消息反序列化失败: {}", e.getMessage());
                }
            }
        }
        
        return messages;
    }
}

/**
 * 延迟消息队列实现
 */
class DelayedMessageQueue {
    
    private final Jedis jedis;
    private final String queueName;
    private final ScheduledExecutorService scheduler;
    
    public DelayedMessageQueue(Jedis jedis, String queueName) {
        this.jedis = jedis;
        this.queueName = queueName;
        this.scheduler = Executors.newScheduledThreadPool(1);
    }
    
    /**
     * 发送延迟消息
     */
    public <T> boolean sendDelayedMessage(T message, long delay, TimeUnit unit) {
        try {
            String serializedMessage = objectMapper.writeValueAsString(message);
            long delayMillis = unit.toMillis(delay);
            long deliverTime = System.currentTimeMillis() + delayMillis;
            
            // 使用有序集合存储延迟消息
            jedis.zadd(queueName + ":delayed", deliverTime, serializedMessage);
            
            // 调度检查任务
            scheduleDeliveryCheck();
            
            return true;
        } catch (JsonProcessingException e) {
            logger.error("消息序列化失败: {}", e.getMessage());
            return false;
        }
    }
    
    /**
     * 调度消息投递检查
     */
    private void scheduleDeliveryCheck() {
        scheduler.schedule(() -> {
            try {
                checkAndDeliverMessages();
            } catch (Exception e) {
                logger.error("延迟消息投递检查失败: {}", e.getMessage());
            }
        }, 1, TimeUnit.SECONDS);
    }
    
    /**
     * 检查并投递到期消息
     */
    private void checkAndDeliverMessages() {
        long currentTime = System.currentTimeMillis();
        
        // 获取所有到期的消息
        Set<String> expiredMessages = jedis.zrangeByScore(
                queueName + ":delayed", 0, currentTime);
        
        if (!expiredMessages.isEmpty()) {
            // 投递到实际队列
            String[] messagesArray = expiredMessages.toArray(new String[0]);
            jedis.rpush(queueName, messagesArray);
            
            // 从延迟队列中移除
            jedis.zremrangeByScore(queueName + ":delayed", 0, currentTime);
        }
        
        // 继续调度下一次检查
        scheduleDeliveryCheck();
    }
    
    /**
     * 接收消息
     */
    public <T> T receiveMessage(Class<T> clazz, int timeoutSeconds) {
        List<String> messages = jedis.blpop(timeoutSeconds, queueName);
        
        if (messages != null && messages.size() >= 2) {
            String message = messages.get(1);
            try {
                return objectMapper.readValue(message, clazz);
            } catch (JsonProcessingException e) {
                logger.error("消息反序列化失败: {}", e.getMessage());
                return null;
            }
        }
        
        return null;
    }
    
    /**
     * 关闭队列
     */
    public void shutdown() {
        scheduler.shutdown();
    }
}

/**
 * 消息队列测试类
 */
class MessageQueueTest {
    
    public static void main(String[] args) throws Exception {
        Jedis jedis = RedisConfig.getJedis();
        
        try {
            System.out.println("=== 测试List消息队列 ===");
            ListMessageQueue listQueue = new ListMessageQueue(jedis, "test:list:queue");
            
            // 发送测试消息
            Map<String, Object> testMessage = Map.of("type", "test", "content", "Hello List Queue");
            listQueue.sendMessage(testMessage);
            
            // 接收消息
            @SuppressWarnings("unchecked")
            Map<String, Object> receivedMessage = (Map<String, Object>) 
                    listQueue.receiveMessage(Object.class, 5);
            System.out.println("List队列接收消息: " + receivedMessage);
            
            System.out.println("\n=== 测试Pub/Sub消息队列 ===");
            PubSubMessageQueue pubSubQueue = new PubSubMessageQueue(jedis, "test:pubsub:channel");
            
            // 订阅消息
            CountDownLatch latch = new CountDownLatch(1);
            pubSubQueue.subscribeMessage(Object.class, message -> {
                System.out.println("Pub/Sub队列接收消息: " + message);
                latch.countDown();
            });
            
            // 发布消息
            Thread.sleep(100); // 等待订阅建立
            Map<String, Object> pubSubMessage = Map.of("type", "test", "content", "Hello Pub/Sub");
            pubSubQueue.publishMessage(pubSubMessage);
            
            // 等待消息接收
            latch.await(5, TimeUnit.SECONDS);
            pubSubQueue.shutdown();
            
            System.out.println("\n=== 测试Stream消息队列 ===");
            StreamMessageQueue streamQueue = new StreamMessageQueue(jedis, "test:stream:queue");
            
            // 发送消息
            Map<String, Object> streamMessage = Map.of("type", "test", "content", "Hello Stream");
            StreamEntryID messageId = streamQueue.sendMessage(streamMessage);
            System.out.println("Stream队列发送消息，ID: " + messageId);
            
            // 读取消息
            List<Object> streamMessages = streamQueue.readMessages(Object.class, 10);
            System.out.println("Stream队列读取消息: " + streamMessages);
            
            System.out.println("\n=== 测试延迟消息队列 ===");
            DelayedMessageQueue delayedQueue = new DelayedMessageQueue(jedis, "test:delayed:queue");
            
            // 发送延迟消息（2秒后投递）
            Map<String, Object> delayedMessage = Map.of("type", "test", "content", "Hello Delayed Queue");
            delayedQueue.sendDelayedMessage(delayedMessage, 2, TimeUnit.SECONDS);
            
            // 立即尝试接收（应该收不到）
            Object immediateMessage = delayedQueue.receiveMessage(Object.class, 1);
            System.out.println("立即接收延迟消息: " + immediateMessage);
            
            // 等待2秒后接收
            Thread.sleep(2500);
            Object delayedReceivedMessage = delayedQueue.receiveMessage(Object.class, 1);
            System.out.println("延迟后接收消息: " + delayedReceivedMessage);
            
            delayedQueue.shutdown();
            
            System.out.println("\n消息队列测试完成!");
            
        } finally {
            RedisConfig.closeJedis(jedis);
        }
    }
}